DietPi-Globals | G_CONFIG_INJECT: Auto escaping of special characters

MichaIng opened the pull request:
**Status**: Testing
- [x] Check all DietPi code internal calls for extended regular expression use
- [x] Check all DietPi code internal calls for escaping within $2
- [x] Check newline support
- [x] Changelog

**Reference**: https://github.com/Fourdee/DietPi/issues/2215

**Commit list/description**:
+ DietPi-Globals | G_CONFIG_INJECT: Minor wording and error catching enhancements
  - Now shows `sed` error output in whiptail prompt as well
+ DietPi-Globals | G_CONFIG_INJECT: Do required special character escaping (**on $2 only!**) function internally
  - Only the usual single and double quote string escaping needs to be done, when not adding via variable.
+ DietPi-Globals | G_CONFIG_INJECT: Remove extended regular expression support
  - The ext regex support doesn't outweigh the way larger escaping effort and bug vulnerability. And even with basic regex only, `(pattern1|pattern2)` or `[char]+` can be done via `\(pattern1\|pattern2\)` resp. `[char]\+`.
+ DietPi-Globals | G_CONFIG_INJECT: Use more individual option variable names

MichaIng committed:
v6.18

+ DietPi-Globals | G_CONFIG_INJECT: Minor wording and error catching enhancements
+ DietPi-Globals | G_CONFIG_INJECT: Do required special character escaping function internally
+ DietPi-Globals | G_CONFIG_INJECT: Remove extended regular expression support

MichaIng commented:
@Fourdee 
If you find time, would be great if you could play around with various special characters and try to cause a syntax error.

Should not be possible any more. If it survives enough tests, I will remove all the error handling from the function. Makes it ugly 😉.

MichaIng committed:
v6.18

+ DietPi-Globals | G_CONFIG_INJECT: Add escaping of forgotten "|", which is used as custom "sed" delimiter

Fourdee commented:
@MichaIng 

Great work 👍 

> If you find time, would be great if you could play around with various special characters and try to cause a syntax error.

Hopefully i'll get some time tomorrow, will run some tests.

MichaIng commented:
@Fourdee 
Sh*t, there is one big issue with automated escaping...

Since all special characters are taken literally, this breaks the use of regex expressions. So, e.g. `G_CONFIG_INJECT 'setting1[[:blank:]]*=' ...` is not possible any more... and there are many cases where settings allow optional-only spaces and other cases where we want to used regex.

Ah have an idea:
- The regex pattern is usually only required for `$1` and optional `$4`, where usually never a variable is inside. So there we can allow regex, skip escaping, thus require manual escaping if some special char is literally wanted.
- The setting `$2` usually contains a variable and we do not need regex there, they would not even be used of course by sed. So there we can do automated escaping.
- It is not 100% consistent: Special chars need to be escaped in $1 and $4 then, but not $2. But for practical use this is the only solution I see that does not require more complicated handling, via e.g. "escaping function" to separately escape variables to pass to G_CONFIG_INJECT first or such...

MichaIng committed:
v6.18

+ DietPi-Config | G_CONFIG_INJECT: Auto-escape $2 settings argument only, to stay flexible with regex on matching pattern arguments

MichaIng committed:
v6.18

+ DietPi-Patch | Remove escaping due to G_CONFIG_INJECT changes

MichaIng committed:
v6.18

+ DietPi-PREP | Remove escaping due to G_CONFIG_INJECT changes

MichaIng committed:
v6.18

+ DietPi-LetsEncrypt | Replace ext regex with basic regex within G_CONFIG_INJECT arguments

MichaIng committed:
v6.18

+ DietPi-Patch | Remove forgotten escaping due to G_CONFIG_INJECT changes

MichaIng committed:
v6.18

+ DietPi-Software | Remove escaping and ext regex due to G_CONFIG_INJECT changes

MichaIng commented:
- [x] Without extended regular expressions, `\|` within grep is taken as `or`, instead of literally. This is the same issue as mentioned above with `\(1\|2\)` which will be interpreted as `1 or 2`. So finally it is better to re-enable extended regular expressions, to allow these characters being interpreted literally at all. Otherwise with unlucky e.g. passwords, such escaped characters could be specially treated 🤣.
```
root@VM-Stretch:/tmp# grep "^[[:blank:]]*setting2 = \|hmmm\|" test
# Added by DietPi:
setting1 = "testing"
setting2 = |hmm|
root@VM-Stretch:/tmp# grep -E "^[[:blank:]]*setting2 = \|hmmm\|" test
root@VM-Stretch:/tmp# grep -E "^[[:blank:]]*setting2 = \|hmm\|" test
setting2 = |hmm|
```
€: 🈯️ Re-enabled ext regex. More escaping, but no other chance to avoid surprises

- [x] Newline is not supported within the setting. This is actually good, since we cannot predict when it is really wanted and when.
- Solution is new option, to explicitly allow newlines. Simply replace `\\n` with `\n` then.

€: 🈯️ Enabled newline support via GCI_NEWLINE (**G**\_**C**ONFIG\_**I**NJECT)

- [x] Whiptail handles newline calls differently. It is not possible to escape them there.
- `whiptail --msgbox 'line1\\nline2' 10 230` echo, grep, sed, I guess POSIX conform is to read the backslash as escaped, thus literally, thus not expand the newline. But whiptail always does 🤔. However, this is minor...

€: Leave as is, very rare and visual only issue, no workaround possible, due to whiptail internal handling

MichaIng committed:
v6.18

+ DietPi-Globals | G_CONFIG_INJECT: Re-enable extended regular expression support, as otherwise \( \| \? \+ are specially treated and not literally...
+ DietPi-Globals | G_CONFIG_INJECT: Use more individual option variable names

MichaIng committed:
v6.18

+ Revert ext regex removal

MichaIng committed:
v6.18

+ Revert ext regex removal + within double quotes \\ still needs to be \\\\

MichaIng committed:
v6.18

+ syntax

MichaIng committed:
v6.18

+ syntax

MichaIng committed:
v6.18

+ syntax

MichaIng committed:
v6.18

+ DietPi-Software | Switch to new G_CONFIG_INJECT option var naming

MichaIng committed:
v6.18

+ Minor

MichaIng committed:
v6.18

+ CHANGELOG | G_CONFIG_INJECT internal special character escaping

Fourdee commented on dietpi/dietpi-software:
```diff
@@ -7897,27 +7897,27 @@ _EOF_
 			if (( ${aSOFTWARE_INSTALL_STATE[83]} >= 1 )); then
 
-				PRESERVE=1 G_CONFIG_INJECT "'htaccess.RewriteBase'" "'htaccess.RewriteBase' => '/owncloud'," $config_php "'overwrite.cli.url'"
+				CGI_PRESERVE=1 G_CONFIG_INJECT "'htaccess.RewriteBase'" "'htaccess.RewriteBase' => '/owncloud'," $config_php "'overwrite.cli.url'"
```
GCI? 😃

MichaIng replied:
@Fourdee
:+1: Whoopsie, fixed

MichaIng committed:
v6.18

typo

Fourdee commented:
@MichaIng 

Thanks Micha, great work 👍 

Merged for testing.
